README.md
<img src="https://assets.armorblox.com/f/52352/775x159/8fa6246e47/logo_color.svg" width=387 alt="Armorblox logo">

# Armorblox Python SDK (Alpha)

[![PyPI version](https://badge.fury.io/py/armorblox-sdk.svg)](https://badge.fury.io/py/armorblox-sdk)
[![Apache-2 License](https://img.shields.io/badge/license-Apache2-blueviolet)](https://www.apache.org/licenses/LICENSE-2.0)

This is an alpha version of the SDK with limited documentation and no support.

## Requirements

Python 3.5+

## Installation

```
pip install armorblox-sdk
```

## Usage

```
from armorblox import client

# Create an API client for your tenant
c = client.Client(api_key='your-api-key-here', instance_name='yourtenantname')

# Fetch information about an incident's analysis data
incident_analysis = c.incidents.analysis(78143)

# Fetch information about an incident's sender data
incident_senders = c.incidents.senders(78143)

# Fetch information about an object associated with an incident (usually mail). 
# Get the object ID from Get Incident by Id's response, under .events[].object_id
incident_object = c.incidents.mail('d72c07bc789c30cb4d63d78ee2861f94add695f9c812e30cfb081b20d3e7e5e7')

# Updates the action to be taken for an incident's objects
update_details = c.incidents.update(78143, body = {
                  "policyActionType": "DELETE",
                  "addSenderToException": False,
                  "actionProfileId": ""
                })


# Fetch a list of threats
threat_incidents, next_page_token, total_incident_count = c.threats.list()

# Fetch a specific threat
incident = c.threats.get(44006)


# Fetch a list of abuse incidents
abuse_incidents, next_page_token, total_incident_count = c.abuse_incidents.list()

# Fetch a specific abuse incident
abuse_incident = c.abuse_incidents.get(44200)


# Fetch a list of DLP incidents
dlp_incidents, next_page_token, total_incident_count = c.dlp_incidents.list()

# Fetch a specific DLP incident
dlp_incident = c.dlp_incidents.get(44010)


# Fetch a list of EAC incidents
eac_incidents, next_page_token, total_incident_count = c.eac_incidents.list()

# Fetch a specific EAC incident
eac_incident = c.eac_incidents.get(67)


# Fetch a list of Graymail incidents
graymail_incidents, next_page_token, total_incident_count = c.graymail_incidents.list()

# Fetch a specific EAC incident
graymail_incident = c.graymail_incidents.get(2627)


# Example to fetch all threats using next_page_token
next_page_token = None
incidents = []
while True:
    threats, next_page_token, total_incident_count = c.threats.list(page_token=next_page_token)
    incidents.extend(threats)
    if not next_page_token:
        break
```

## Contributing

* Install [Poetry](https://python-poetry.org)
* Clone the SDK repo & `cd` into it
```
git clone https://github.com/armorblox/armorblox-python-sdk
cd armorblox-python-sdk
```
* Run `poetry install` to install the dependencies
* Run `tox` to run the tests

## Publishing

#### TestPyPI

One-time setup
```
poetry config repositories.test-pypi https://test.pypi.org/legacy/
poetry config pypi-token.test-pypi <your-TestPyPI-token>
```

Publishing
```
poetry publish --build -r test-pypi
```

Use
```
pip install --index-url https://test.pypi.org/simple/ --no-deps armorblox-sdk
```
to make sure the installation works correctly.

#### PyPI

One-time setup
```
poetry config pypi-token.pypi <your-PyPI-token>
```

Publishing
```
poetry publish --build
```


CHANGELOG.md
# Change Log

## Version 0.1.4 (2023-03-20)

* Support for EAC and Graymail incident types

## Version 0.1.3 (2022-11-16)

* Support for incident analysis, senders, mails & update action APIs
* Bugfixes & performance fixes
* README updates

## Version 0.1.2 (2022-07-26)

* Support for returning page token and total incident count in the incident list method

## Version 0.1.1 (2022-07-22)

* Add support for passing in `params` in incident list method

## Version 0.1.0 (2022-07-01)

* Add basic support for incident list and detail APIs


tests/api/test_incidents_api.py
# -*- coding: utf-8 -*-


from armorblox.api.incidents_api import IncidentsApi, ThreatsApi, DLPIncidentsApi, AbuseIncidentsApi, EACIncidentsApi, GraymailIncidentsApi
from armorblox.config import Config
import pytest
import responses


class TestIncidentsApi:
    @pytest.fixture(autouse=True)
    def set_incidents_api(self, config_params):
        self.config = Config(**config_params)
        self.base_api_url = self.config.base_api_url.format('v1beta1')
        self.incidents_api = IncidentsApi(self.config)

    def test_list(self, mocked_responses):
        mocked_responses.add(
            responses.GET, self.base_api_url + IncidentsApi.PATH,
            json={
                'total_count': 150,
                'next_page_token': '50',
                'incidents': [{'a': 1}]
            },
            status=200,
            content_type='application/json')
        incidents, next_page_token, total_count = self.incidents_api.list()
        assert total_count == 150
        assert next_page_token == '50'


src/armorblox/client.py
# -*- coding: utf-8 -*-


from armorblox.config import Config
from armorblox.api.incidents_api import IncidentsApi, ThreatsApi, DLPIncidentsApi, AbuseIncidentsApi, EACIncidentsApi, GraymailIncidentsApi
from cached_property import cached_property


class Client:

    @staticmethod
    def sdk_version():
        return '0.1.0.20220201'

    def __init__(self, api_key: str, instance_name: str = None, instance_url: str = None):
        self.config = Config(api_key,
                             instance_name=instance_name,
                             instance_url=instance_url,
                             sdk_version=self.sdk_version())

    @cached_property
    def incidents(self):
        return IncidentsApi(self.config)

    @cached_property
    def threats(self):
        return ThreatsApi(self.config)

    @cached_property
    def dlp_incidents(self):
        return DLPIncidentsApi(self.config)

    @cached_property
    def abuse_incidents(self):
        return AbuseIncidentsApi(self.config)

    @cached_property
    def eac_incidents(self):
        return EACIncidentsApi(self.config)
    
    @cached_property
    def graymail_incidents(self):
        return GraymailIncidentsApi(self.config)

src/armorblox/config.py
# -*- coding: utf-8 -*-


from urllib.parse import urlparse


class Config:
    BASE_INSTANCE_URL = 'https://{}.armorblox.io/'
    BASE_API_PATH = 'api/{}/organizations/{}/'

    def __init__(
            self, api_key: str, instance_name: str = None, instance_url: str = None,
            sdk_version: str = 'Unspecified', incident_id: int = None):
        """

        Args:
            api_key: string
            instance_name: string
            instance_url: string
            sdk_version: string

        Raises:
            Exception:
        """
        self._sdk_version = sdk_version
        self._api_key = api_key
        self._incident_id = incident_id

        if instance_url:
            self._instance_url = instance_url
            if not instance_url.endswith('/'):
                self._instance_url += '/'
            if instance_name:
                self._instance_name = instance_name
            else:
                self._instance_name = urlparse(instance_url).netloc.split('.')[0]
        elif instance_name:
            self._instance_url = self.BASE_INSTANCE_URL.format(instance_name)
            self._instance_name = instance_name
        else:
            raise Exception('At least one of instance_url or instance_name should be provided')

        self._base_api_url = self._instance_url + self.BASE_API_PATH.format('{}', self._instance_name)


    @property
    def sdk_version(self):
        return self._sdk_version

    @property
    def api_key(self):
        return self._api_key

    @property
    def instance_url(self):
        return self._instance_url

    @property
    def instance_name(self):
        return self._instance_name

    @property
    def base_api_url(self):
        return self._base_api_url

src/armorblox/api/incidents_api.py
# -*- coding: utf-8 -*-


from armorblox.api.base_api import BaseApi
from enum import Enum, unique


@unique
class IncidentType(Enum):
    THREAT_INCIDENT_TYPE = 1
    DLP_INCIDENT_TYPE = 2
    ABUSE_INCIDENT_TYPE = 3
    EAC_INCIDENT_TYPE = 4
    GRAYMAIL_INCIDENT_TYPE = 5


class IncidentsApi(BaseApi):
    PATH = 'incidents'
    TIME_FORMAT = '%Y-%m-%dT%H:%M:%SZ'

    def __init__(self, config, incident_type=None):
        super().__init__(config)
        self._incident_type = incident_type

    def list(self, page_token=None, params=None):
        if params is None:
            params = {
                'orderBy': 'DESC',
                'sortBy': 'DATE',
                'timeFilter': 'allTime'
            }
        if self._incident_type is not None:
            params['incidentTypesFilter'] = self._incident_type.name
        if page_token is not None:
            params['page_token'] = page_token

        response_json, response = self.list_resource(self.PATH, params=params)
        if response_json is None:
            return [], None, 0
        else:
            next_page_token = response_json.get('next_page_token', None)
            total_count = int(response_json.get('total_count', 0))
            incidents = response_json.get('incidents', [])
            return incidents, next_page_token, total_count

    def get(self, incident_id: int):
        params = {
            'pageSize': 20
        }

        return self.get_resource(self.PATH, str(incident_id), params=params)

    def analysis(self, incident_id: int):
        return self.get_resource(self.PATH, str(incident_id) + ':getAnalysis')

    def senders(self, incident_id: int):
        return self.get_resource(self.PATH, str(incident_id) + ':getSenders')

    def mail(self, object_id: str):
        params = {'objectId': object_id, 'objectType':'CONTENT_MAIL'}
        options = {'api_version': 'v1alpha1'}
        return self.get_resource('', ':getPolicyViolationDetailsByObjectId',
                                 params=params, options=options)

    def update(self, incident_id: int, body: dict):
        return self.update_resource(self.PATH, str(incident_id) + ':updateAction',
                                    body=body)


class ThreatsApi(IncidentsApi):
    def __init__(self, config):
        super().__init__(config, IncidentType.THREAT_INCIDENT_TYPE)


class DLPIncidentsApi(IncidentsApi):
    def __init__(self, config):
        super().__init__(config, IncidentType.DLP_INCIDENT_TYPE)


class AbuseIncidentsApi(IncidentsApi):
    def __init__(self, config):
        super().__init__(config, IncidentType.ABUSE_INCIDENT_TYPE)


class EACIncidentsApi(IncidentsApi):
    def __init__(self, config):
        super().__init__(config, IncidentType.EAC_INCIDENT_TYPE)


class GraymailIncidentsApi(IncidentsApi):
    def __init__(self, config):
        super().__init__(config, IncidentType.GRAYMAIL_INCIDENT_TYPE)

src/armorblox/api/base_api.py
# -*- coding: utf-8 -*-


import requests
import json

class BaseApi:
    PAGE_SIZE = 50

    def __init__(self, config):
        self._config = config

    @property
    def version(self):
        return 'v1beta1'

    def headers(self):
        return {
            'X-Ab-Authorization': self._config.api_key,
            'Content-Type': 'application/json',
            'User-Agent': 'Armorblox-Python-SDK/' + self._config.sdk_version
        }

    def endpoint(self, path: str, api_version: str = None) -> str:
        if api_version is None:
            api_version = self.version
        return self._config.base_api_url.format(api_version) + path

    def list_params(self):
        return {
            'page_size': self.PAGE_SIZE,
            'page_token': '0',
        }

    def list_resource(self, path: str, headers: dict = None,
                      params: dict = None, options: dict = None):
        """

        Args:
            path: str
            headers: dict
            params: dict
            options: dict

        Raises:
            Exception:
        """
        h = self.headers()
        if headers is not None:
            h.update(headers)
        p = self.list_params()
        if params is not None:
            p.update(params)
            # Cap page size to PAGE_SIZE value for optimal performance
            p["page_size"] = min(p["page_size"], self.PAGE_SIZE)
        if options is None:
            options = {}
        response = requests.get(self.endpoint(path, options.get('api_version')),
                                headers=h, params=p)
        if response.status_code == 200:
            return response.json(), None
        else:
            return None, response
    
    def get_resource(self, path: str, resource_id: str, headers: dict = None,
                     params: dict = None, options: dict = None):
        """

        Args:
            path: str
            resource_id: str
            headers: dict
            params: dict
            options: dict
        Raises:
            Exception:
        """
        h = self.headers()
        if headers is not None:
            h.update(headers)
        if options is None:
            options = {}

        url = self.endpoint(path, options.get('api_version'))
        if resource_id.startswith(':'):
            url = f"{url.rstrip('/')}{resource_id}"
        else:
            url = f"{url}/{resource_id}"

        response = requests.get(url, headers=h, params=params)

        if response.status_code == 200:
            return response.json()
        else:
            return {}


    def update_resource(self, path: str, resource_id: str, headers: dict = None,
                     params: dict = None, options: dict = None, body: dict = None):
        """

        Args:
            path: str
            resource_id: str
            headers: dict
            params: dict
            options: dict
            body: dict

        Raises:
            Exception:
        """
        h = self.headers()
        if headers is not None:
            h.update(headers)
        if options is None:
            options = {}

        url = self.endpoint(path, options.get('api_version')) + f"/{resource_id}" + ":updateAction"
        response = requests.request("PATCH", url, headers=h, params=params, data=json.dumps(body))
        if response.status_code == 200:
            return response.json()
        else:
            return {}